### ISSUE TITLE
Cache catch-all bug

### ISSUE THREAD

**Issue by gpsaggese** [alphamatic/amp#1675]
- [ ] Run coverage for cache and make sure everything is covered
  ```
  > i run_fast_slow_tests --pytest-opts="helpers/test/test_cache.py" --coverage
  ```

- [ ] When there is a disk cache miss, also the memory cache should be populated, Instead now we wait for another cache miss
   - WONTFIX?

- [X] Have a way to update the saved code
   - https://app.zenhub.com/workspaces/alphamatic-6083f954eb3330000e834e01/issues/alphamatic/amp/1665

- [X] Check if a function will be executed from a cache or not. Instrument the code and throw an exception if the function will be recomputed (instead of recomputing)
- [X] Have read-only caches
- [ ] Add way to disable caching on a function basis

- [ ] When running unit tests allow to disable caching
  - https://github.com/alphamatic/amp/issues/765
  - https://stackoverflow.com/questions/53318600/disable-joblib-memory-caching-globally-during-unittest
  - https://app.zenhub.com/workspaces/alphamatic-6083f954eb3330000e834e01/issues/alphamatic/amp/765
  - The problem here is that some tests need the cache to be enabled (i.e., the tests for the cache)
  - One solution is to just add a switch to `git_clean` to clean the client and / or to clean the unit test cache dir before starting the tests

- [ ] Make sure we don't call du (through info) in every call 

- [ ] When the cache is read-only (and maybe even if it's not), check that the dst dir exists

**Comment by gpsaggese** [comment]
- [ ] Propagate support for ignore
  - See https://joblib.readthedocs.io/en/latest/memory.html#ignoring-some-arguments, e.g., `ignore=['debug']`
  - We want to pass ignore to _Cached ctor and then pass it to `memory_cache.cache(self._func)` in `_create_function_memory_cache`
  - It's unclear if `_has_cached_version` supports the `ignore`
